File: CS/CodeExamples/CommentsActions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using DevExpress.XtraRichEdit.API.Native;
using DevExpress.XtraRichEdit;
using System.Diagnostics;

namespace RichEditDocumentServerAPIExample.CodeExamples
{
    public class CommentsActions
    {
        public static Action<RichEditDocumentServer> CreateCommentAction = CreateComment;
        public static Action<RichEditDocumentServer> CreateNestedCommentAction = CreateNestedComment;
        public static Action<RichEditDocumentServer> DeleteCommentAction = DeleteComment;
        public static Action<RichEditDocumentServer> EditCommentPropertiesAction = EditCommentProperties;
        public static Action<RichEditDocumentServer> EditCommentContentAction = EditCommentContent;

        static void CreateComment(RichEditDocumentServer wordProcessor)
        {
            #region #CreateComment
            // Load a document from a file.
            wordProcessor.LoadDocument("Documents\\Grimm.docx", DocumentFormat.OpenXml);

            // Access a document.
            Document document = wordProcessor.Document;

            if (document.Paragraphs.Count > 2)
            {
                // Access the range of the third paragraph.
                DocumentRange docRange = document.Paragraphs[2].Range;

                // Specify the comment's author name.
                string commentAuthor = "Johnson Alphonso D";

                // Create a comment.
                document.Comments.Create(docRange, commentAuthor, DateTime.Now);
            }
            #endregion #CreateComment
        }

        static void CreateNestedComment(RichEditDocumentServer wordProcessor)
        {
            #region #CreateNestedComment
            // Load a document from a file.
            wordProcessor.LoadDocument("Documents\\Grimm.docx", DocumentFormat.OpenXml);

            // Access a document.
            Document document = wordProcessor.Document;

            if (document.Comments.Count > 1)
            {
                // Find text ranges matched the string in the document range
                // to which the parent comment relates.
                DocumentRange[] resRanges = document.FindAll("trump", SearchOptions.None, document.Comments[1].Range);
                if (resRanges.Length > 0)
                {
                    // Create a new comment nested in the parent comment.
                    Comment newComment = document.Comments.Create("Vicars Anny", document.Comments[1]);
                    newComment.Date = DateTime.Now;
                }
            }
            #endregion #CreateNestedComment
        }

        static void DeleteComment(RichEditDocumentServer wordProcessor)
        {
            #region #DeleteComment
            // Load a document from a file.
            wordProcessor.LoadDocument("Documents\\Grimm.docx", DocumentFormat.OpenXml);

            // Access a document.
            Document document = wordProcessor.Document;

            if (document.Comments.Count > 0)
            {
                // Delete the first comment.
                document.Comments.Remove(document.Comments[0]);
            }
            #endregion #DeleteComment
        }

        static void EditCommentProperties(RichEditDocumentServer wordProcessor)
        {
            #region #EditCommentProperties
            // Load a document from a file.
            wordProcessor.LoadDocument("Documents\\Grimm.docx", DocumentFormat.OpenXml);

            // Access a document.
            Document document = wordProcessor.Document;

            int commentCount = document.Comments.Count;
            if (commentCount > 0)
            {
                // Start to edit the document.
                document.BeginUpdate();

                // Access a comment and edit its properties.
                Comment comment = document.Comments[document.Comments.Count - 1];
                comment.Name = "New Name";
                comment.Date = DateTime.Now;
                comment.Author = "New Author";
                
                // Finalize to edit the document.
                document.EndUpdate();
            }
            #endregion #EditCommentProperties
        }

        static void EditCommentContent(RichEditDocumentServer wordProcessor)
        {
            #region #EditCommentContent
            // Load a document from a file.
            wordProcessor.LoadDocument("Documents\\Grimm.docx", DocumentFormat.OpenXml);

            // Access a document.
            Document document = wordProcessor.Document;

            int commentCount = document.Comments.Count;
            if (commentCount > 0)
            {
                // Access a comment.
                Comment comment = document.Comments[document.Comments.Count - 1];
                if (comment != null)
                {
                    // Start to edit the comment.
                    SubDocument commentDocument = comment.BeginUpdate();

                    // Insert a text to the comment.
                    commentDocument.InsertText(commentDocument.CreatePosition(0), "some text");

                    // Insert a table to the comment.
                    commentDocument.Tables.Create(commentDocument.CreatePosition(9), 5, 4);

                    // Finalize to edit the comment.
                    comment.EndUpdate(commentDocument);
                }
            }
            #endregion #EditCommentContent
        }


    }
}

File: VB/CodeExamples/CommentsActions.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports DevExpress.XtraRichEdit.API.Native
Imports DevExpress.XtraRichEdit
Imports System.Diagnostics

Namespace RichEditDocumentServerAPIExample.CodeExamples

    Public Class CommentsActions

        Public Shared CreateCommentAction As System.Action(Of DevExpress.XtraRichEdit.RichEditDocumentServer) = AddressOf RichEditDocumentServerAPIExample.CodeExamples.CommentsActions.CreateComment

        Public Shared CreateNestedCommentAction As System.Action(Of DevExpress.XtraRichEdit.RichEditDocumentServer) = AddressOf RichEditDocumentServerAPIExample.CodeExamples.CommentsActions.CreateNestedComment

        Public Shared DeleteCommentAction As System.Action(Of DevExpress.XtraRichEdit.RichEditDocumentServer) = AddressOf RichEditDocumentServerAPIExample.CodeExamples.CommentsActions.DeleteComment

        Public Shared EditCommentPropertiesAction As System.Action(Of DevExpress.XtraRichEdit.RichEditDocumentServer) = AddressOf RichEditDocumentServerAPIExample.CodeExamples.CommentsActions.EditCommentProperties

        Public Shared EditCommentContentAction As System.Action(Of DevExpress.XtraRichEdit.RichEditDocumentServer) = AddressOf RichEditDocumentServerAPIExample.CodeExamples.CommentsActions.EditCommentContent

        Private Shared Sub CreateComment(ByVal wordProcessor As DevExpress.XtraRichEdit.RichEditDocumentServer)
#Region "#CreateComment"
            ' Load a document from a file.
            wordProcessor.LoadDocument("Documents\Grimm.docx", DevExpress.XtraRichEdit.DocumentFormat.OpenXml)
            ' Access a document.
            Dim document As DevExpress.XtraRichEdit.API.Native.Document = wordProcessor.Document
            If document.Paragraphs.Count > 2 Then
                ' Access the range of the third paragraph.
                Dim docRange As DevExpress.XtraRichEdit.API.Native.DocumentRange = document.Paragraphs(CInt((2))).Range
                ' Specify the comment's author name.
                Dim commentAuthor As String = "Johnson Alphonso D"
                ' Create a comment.
                document.Comments.Create(docRange, commentAuthor, System.DateTime.Now)
            End If
#End Region  ' #CreateComment
        End Sub

        Private Shared Sub CreateNestedComment(ByVal wordProcessor As DevExpress.XtraRichEdit.RichEditDocumentServer)
#Region "#CreateNestedComment"
            ' Load a document from a file.
            wordProcessor.LoadDocument("Documents\Grimm.docx", DevExpress.XtraRichEdit.DocumentFormat.OpenXml)
            ' Access a document.
            Dim document As DevExpress.XtraRichEdit.API.Native.Document = wordProcessor.Document
            If document.Comments.Count > 1 Then
                ' Create a new comment nested in the parent comment.
                Dim newComment As DevExpress.XtraRichEdit.API.Native.Comment = document.Comments.Create("Vicars Anny", document.Comments(1))
                newComment.[Date] = System.DateTime.Now
                Dim commentDocument As SubDocument = newComment.BeginUpdate()
                commentDocument.InsertText(commentDocument.Range.Start, "I agree")
                newComment.EndUpdate(commentDocument)
            End If
#End Region  ' #CreateNestedComment
        End Sub

        Private Shared Sub DeleteComment(ByVal wordProcessor As DevExpress.XtraRichEdit.RichEditDocumentServer)
#Region "#DeleteComment"
            ' Load a document from a file.
            wordProcessor.LoadDocument("Documents\Grimm.docx", DevExpress.XtraRichEdit.DocumentFormat.OpenXml)
            ' Access a document.
            Dim document As DevExpress.XtraRichEdit.API.Native.Document = wordProcessor.Document
            If document.Comments.Count > 0 Then
                ' Delete the first comment.
                document.Comments.Remove(document.Comments(0))
            End If
#End Region  ' #DeleteComment
        End Sub

        Private Shared Sub EditCommentProperties(ByVal wordProcessor As DevExpress.XtraRichEdit.RichEditDocumentServer)
#Region "#EditCommentProperties"
            ' Load a document from a file.
            wordProcessor.LoadDocument("Documents\Grimm.docx", DevExpress.XtraRichEdit.DocumentFormat.OpenXml)
            ' Access a document.
            Dim document As DevExpress.XtraRichEdit.API.Native.Document = wordProcessor.Document
            Dim commentCount As Integer = document.Comments.Count
            If commentCount > 0 Then
                ' Start to edit the document.
                document.BeginUpdate()
                ' Access a comment and edit its properties.
                Dim comment As DevExpress.XtraRichEdit.API.Native.Comment = document.Comments(document.Comments.Count - 1)
                comment.Name = "New Name"
                comment.[Date] = System.DateTime.Now
                comment.Author = "New Author"
                ' Finalize to edit the document.
                document.EndUpdate()
            End If
#End Region  ' #EditCommentProperties
        End Sub

        Private Shared Sub EditCommentContent(ByVal wordProcessor As DevExpress.XtraRichEdit.RichEditDocumentServer)
#Region "#EditCommentContent"
            ' Load a document from a file.
            wordProcessor.LoadDocument("Documents\Grimm.docx", DevExpress.XtraRichEdit.DocumentFormat.OpenXml)
            ' Access a document.
            Dim document As DevExpress.XtraRichEdit.API.Native.Document = wordProcessor.Document
            Dim commentCount As Integer = document.Comments.Count
            If commentCount > 0 Then
                ' Access a comment.
                Dim comment As DevExpress.XtraRichEdit.API.Native.Comment = document.Comments(document.Comments.Count - 1)
                If comment IsNot Nothing Then
                    ' Start to edit the comment.
                    Dim commentDocument As DevExpress.XtraRichEdit.API.Native.SubDocument = comment.BeginUpdate()
                    ' Insert a text to the comment.
                    commentDocument.Paragraphs.Insert(commentDocument.Range.Start)
                    commentDocument.InsertText(commentDocument.Range.Start, "some text")
                    ' Insert a table to the comment.
                    commentDocument.Tables.Create(commentDocument.Range.End, 5, 4)
                    ' Finalize to edit the comment.
                    comment.EndUpdate(commentDocument)
                End If
            End If
#End Region  ' #EditCommentContent
        End Sub
    End Class
End Namespace
